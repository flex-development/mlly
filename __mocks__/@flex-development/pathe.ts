/**
 * @file Mocks - @flex-development/pathe
 * @module mocks/flex-development/pathe
 * @see https://github.com/flex-development/pathe
 */

/**
 * [`@flex-development/pathe`][1] module type.
 *
 * [1]: https://github.com/flex-development/pathe
 */
type Actual = typeof import('@flex-development/pathe')

/**
 * `@flex-development/pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/pathe')

export default {
  basename: vi.fn(actual.basename),
  changeExt: vi.fn(actual.changeExt),
  dirname: vi.fn(actual.dirname),
  extname: vi.fn(actual.extname),
  format: vi.fn(actual.format),
  isAbsolute: vi.fn(actual.isAbsolute),
  join: vi.fn(actual.join),
  parse: vi.fn(actual.parse),
  relative: vi.fn(actual.relative),
  removeExt: vi.fn(actual.removeExt),
  resolve: vi.fn(actual.resolve)
}
