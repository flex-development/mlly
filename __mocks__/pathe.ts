/**
 * @file Mocks - pathe
 * @module mocks/pathe
 * @see https://github.com/unjs/pathe
 */

/**
 * [`pathe`][1] module type.
 *
 * [1]: https://github.com/unjs/pathe
 */
type Actual = typeof import('pathe')

/**
 * `pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('pathe')

export const basename = vi.fn(actual.basename)
export const dirname = vi.fn(actual.dirname)
export const extname = vi.fn(actual.extname)
export const isAbsolute = vi.fn(actual.isAbsolute)
export const join = vi.fn(actual.join)
export const parse = vi.fn(actual.parse)
export const relative = vi.fn(actual.relative)
export const resolve = vi.fn(actual.resolve)
