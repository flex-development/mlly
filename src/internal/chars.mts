/**
 * @file Internal - chars
 * @module mlly/internal/chars
 */

/**
 * Characters that may have special meaning.
 *
 * @enum {string}
 */
enum chars {
  asterisk = '*',
  at = '@',
  empty = '',
  hash = '#'
}

export default chars
