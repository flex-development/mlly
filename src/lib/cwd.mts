/**
 * @file cwd
 * @module mlly/lib/cwd
 */

import pathe from '@flex-development/pathe'

/**
 * Get the URL of the current working directory.
 *
 * @return {URL}
 *  URL of current working directory
 */
function cwd(): URL {
  return pathe.pathToFileURL(pathe.cwd() + pathe.sep)
}

export default cwd
