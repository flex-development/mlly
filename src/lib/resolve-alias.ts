/**
 * @file resolveAlias
 * @module mlly/lib/resolveAlias
 */

import chars from '#internal/chars'
import cwd from '#lib/cwd'
import isRelativeSpecifier from '#lib/is-relative-specifier'
import patternMatch from '#lib/pattern-match'
import toRelativeSpecifier from '#lib/to-relative-specifier'
import type { PatternMatch, ResolveAliasOptions } from '@flex-development/mlly'
import pathe from '@flex-development/pathe'

export default resolveAlias

/**
 * Resolve an aliased `specifier`.
 *
 * @see {@linkcode ResolveAliasOptions}
 *
 * @param {string} specifier
 *  The module specifier to resolve
 * @param {ResolveAliasOptions | null | undefined} [options]
 *  Alias resolution options
 * @return {string | null }
 *  Specifier of aliased module or `null` if path alias match is not found
 */
function resolveAlias(
  specifier: string,
  options?: ResolveAliasOptions | null | undefined
): string | null {
  /**
   * Resolved alias.
   *
   * @var {string | null} aliased
   */
  let aliased: string | null = null

  if (options) {
    if (typeof options.aliases === 'object' && options.aliases) {
      /**
       * Tuple containing expansion key and pattern match.
       *
       * @const {PatternMatch | null} match
       */
      const match: PatternMatch | null = patternMatch(
        specifier,
        options.aliases
      )

      if (match) {
        aliased = resolveAliasTarget(options.aliases[match[0]], match[1])

        if (typeof aliased === 'string') {
          /**
           * URL of aliased module.
           *
           * @const {URL} url
           */
          const url: URL = new URL(aliased, options.cwd ?? cwd())

          if (options.absolute) {
            aliased = url.href
          } else if (options.parent) {
            aliased = toRelativeSpecifier(url, options.parent)
          }
        }
      }
    }
  }

  return aliased
}

/**
 * Resolve a path alias target.
 *
 * @internal
 *
 * @param {unknown} target
 *  The path alias target to resolve
 * @param {string | null | undefined} [patternMatch]
 *  Pattern match, replaces `*` in `target`
 * @return {string | null}
 *  Matched alias
 */
function resolveAliasTarget(
  target: unknown,
  patternMatch?: string | null | undefined
): string | null {
  if (typeof target === 'string') {
    /**
     * Matched alias.
     *
     * @var {string} match
     */
    let match: string = typeof patternMatch === 'string'
      ? target.replace(chars.asterisk, patternMatch)
      : target

    if (!isRelativeSpecifier(match)) match = pathe.dot + pathe.sep + match
    return match
  }

  if (Array.isArray<string | null | undefined>(target)) {
    for (const targetValue of target) {
      /**
       * Resolved path alias.
       *
       * @const {string | null} resolved
       */
      const resolved: string | null = resolveAliasTarget(
        targetValue,
        patternMatch
      )

      if (typeof resolved === 'string') return resolved
    }
  }

  return null
}
