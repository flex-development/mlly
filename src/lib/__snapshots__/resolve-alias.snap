// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:lib/resolveAlias > should return specifier of aliased module ("#lib") 1`] = `"./src/lib/index"`;

exports[`unit:lib/resolveAlias > should return specifier of aliased module ("#lib/resolve-alias") 1`] = `"../resolve-alias.ts"`;

exports[`unit:lib/resolveAlias > should return specifier of aliased module ("#tests/reporters/notifier.ts") 1`] = `file://\${process.cwd()}/__tests__/reporters/notifier.ts`;

exports[`unit:lib/resolveAlias > should return specifier of aliased module ("#tests/utils/to-package-url") 1`] = `"./__tests__/utils/to-package-url"`;
